_integration/testsuite/httpproxy/013-backend-tls.yaml

```diff
@@ -0,0 +1,262 @@
+    - testsuite
+  commonName: issuer
+  secretName: ca-projectcontour-io
+  issuerRef:
+    name: selfsigned
+
+---
+
+apiVersion: cert-manager.io/v1alpha2
+kind: Issuer
+metadata:
+  name: ca-projectcontour-io
+  namespace: 013-backend-tls
+spec:
+  ca:
+    secretName: ca-projectcontour-io
+
+---
+
+apiVersion: cert-manager.io/v1alpha2
+kind: Certificate
+metadata:
+  name: backend-server-cert
+  namespace: 013-backend-tls
+spec:
+  commonName: echo
+  usages:
+  - server auth
+  dnsNames:
+  - echo
+  secretName: backend-server-cert
+  issuerRef:
+    name: ca-projectcontour-io
+
+---
+
+apiVersion: cert-manager.io/v1alpha2
+kind: Certificate
+metadata:
+  name: envoy-client-cert
+  namespace: 013-backend-tls
+spec:
+  commonName: envoy
+  usages:
+  - client auth
+  secretName: envoy-client-cert
+  issuerRef:
+    name: ca-projectcontour-io
+
+---
+
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: ingress-conformance-echo-tls
+$apply:
+  fixture:
+    as: 013-backend-tls/echo
+
+---
+
+apiVersion: v1
+kind: Service
+metadata:
+  name: ingress-conformance-echo-tls
+$apply:
+  fixture:
+    as: 013-backend-tls/echo
+
+---
+
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: echo
+  namespace: 013-backend-tls
+spec:
+  virtualhost:
+    fqdn: echo.projectcontour.io
+  routes:
+  - services:
+    - name: echo
+      port: 443
+      protocol: tls
+      validation:
+        caSecret: envoy-client-cert  # only ca.crt is used from the secret
+        subjectName: echo
+
+---
+
+import data.contour.http.client
+import data.contour.http.client.url
+import data.contour.http.response
+import data.contour.resources
+
+# request is made with HTTP, but the test case verifies that HTTPS is used between envoy and backend
+Response := client.Get({
+  "url": url.http("/echo"),
+  "headers": {
+    "Host": "echo.projectcontour.io",
+    "User-Agent": client.ua("backend-tls-test"),
+  },
+})
+
+EnvoySecret := resources.get("secrets", "013-backend-tls/envoy-client-cert")
+EnvoyCert := base64.decode(EnvoySecret.data["tls.crt"])
+
+error_non_200_response [msg] {
+  not Response
+  msg := "no response"
+}
+
+error_non_200_response [msg] {
+  Response.status_code != 200
+  msg := sprintf("got status %d, wanted %d", [Response.status_code, 200])
+}
+
+error_no_upstream_tls [msg] {
+  not Response.body.tls
+  msg := "Backend did not receive TLS connection from Envoy"
+}
+
+error_no_upstream_clientcert [msg] {
+  not Response.body.tls.client_certificate
+  msg := "Backend did not receive client certificate from Envoy"
+}
+
+error_wrong_tls_certificate [msg] {
+  Response.body.tls.client_certificate != EnvoyCert
+  msg := sprintf("Backend received wrong certificate\n- %s\n+ %s", [EnvoyCert, Response.body.tls.client_certificate])
+}
+
+---
+
+# delete envoy certificate to force cert-manager to rotate it
+apiVersion: v1
+kind: Secret
+metadata:
+  name: envoy-client-cert
+  namespace: 013-backend-tls
+$apply: delete
+
+---
+
```

Comment:
Each fragment gets its own timeout, so the test ends up being more deterministic if you test what you expect in each stage in a separate fragment. Even with this, I think it's possible for the check to spuriously succeed since deleted objects can be visible through the API for a short time.